File: grad/h2o_stv.c
#include <stdio.h>
#include <stdlib.h>
#include <math.h>
#include "cint.h"

int cint1e_ovlp_sph(double *buf, int *shls,
                    int *atm, int natm, int *bas, int nbas, double *env);

int cint1e_kin_cart(double *buf, int *shls,
                    int *atm, int natm, int *bas, int nbas, double *env);

int cint1e_nuc_cart(double *buf, int *shls,
                    int *atm, int natm, int *bas, int nbas, double *env);

typedef int (* Int_cart)(double *buf, int *shls, 
					int *atm, int natm, int *bas, int nbas, double *env);

void read_arrays(FILE * file, int natm, int nbas, int ** atm, int ** bas, double ** env)
{
	double num;
	int number;
	int count = 0;

	while (count < natm * ATM_SLOTS) {
        fscanf(file, "%d", &number);
		(*atm)[count] = number;
		count++;
	}

	// for (int i = 0; i < count; i++) {
	// 	printf("%d ", (*atm)[i]);
	// }
	// printf("\n");

	count = 0;
	while (count < nbas * BAS_SLOTS) {
        fscanf(file, "%d", &number);
		(*bas)[count] = number;
		count++;
	}

	// for (int i = 0; i < count; i++) {
	// 	printf("%d ", (*bas)[i]);
	// }
	// printf("\n");

	count = 0;
	while (fscanf(file, "%lf", &num) != EOF) {
		(*env)[count] = num;
		count++;
	}
	fclose(file);

	// for (int i = 0; i < count; i++) {
	// 	printf("%lf ", (*env)[i]);
	// }
	// printf("\n");
}

int main(int argc, char ** argv)
{	
	int natm = 3;
	int nbas = 7; // 12 for def2svp

	int * atm = malloc(sizeof(int) * natm * ATM_SLOTS);
	int * bas = malloc(sizeof(int) * nbas * BAS_SLOTS);
	double * env = malloc(sizeof(double) * 10000);

	FILE * file = fopen("/u/jpmedina/libcint/molecules/h2o/sto3g.txt", "r");
 	read_arrays(file, natm, nbas, &atm, &bas, &env);
	
	int di, dj;
	int shls[4];

	double * buf;
	
	printf("buf:\n");

	Int_cart func = cint1e_ovlp_sph;

	for (int i = 0; i < nbas; i++) {
		for (int j = 0; j < nbas; j++) {
			shls[0] = i; di = CINTcgto_spheric(i, bas);
			shls[1] = j; dj = CINTcgto_spheric(j, bas);

			// printf("i: %d j: %d di: %d dj %d\n", i, j, di, dj);

			buf = malloc(sizeof(double) * di * dj);

			func(buf, shls, atm, natm, bas, nbas, env);

			for (int k = 0; k < di * dj; k++) {
				printf("%.3f ", buf[k]);
			}
			free(buf);
			// printf("\n");
		}
		printf("\n");
	}
	printf("\n");

	free(atm);
	free(bas);
	free(env);
}

File: librint/grad/h2o/sto3g/h2o_stv.rs
extern crate librint;

use librint::cint_bas::CINTcgto_spheric;
use librint::cint1e::cint1e_ovlp_sph;

pub const ATM_SLOTS: usize = 6;
pub const BAS_SLOTS: usize = 8;

fn int(
    natm: usize,
    nbas: usize,
    nshells: usize,
    atm: &mut [i32],
    bas: &mut [i32],
    env: &mut [f64],
) -> Vec<f64> {
    let mut buf: Vec<f64>;
    let mut shls: [i32; 4] = [0, 0, 0, 0];

    let mut S = vec![0.0; nshells * nshells];

    println!("ovlp");
    let mut mu: usize = 0;
    for i in 0..nbas {
        let mut nu: usize = 0;
        let mut di: usize = 0;
        for j in 0..nbas {
            shls[0] = i as i32;
            shls[1] = j as i32;
            
            di = CINTcgto_spheric(i, &bas) as usize;
            let dj = CINTcgto_spheric(j, &bas) as usize;

            buf = vec![0.0; di * dj];
            cint1e_ovlp_sph(&mut buf, &mut shls, atm, natm as i32, bas, nbas as i32, env);

            // Add the buf in the correct place
            let mut c: usize = 0;
            for mui in mu..(mu + di) {
                for nuj in nu..(nu + dj) {
                    S[mui * nshells + nuj] = buf[c];
                    c += 1;
                }
            }
            nu += dj;
        }
        mu += di;
    }

    return S;
}

fn main() {
	const natm: usize = 3;
	const nbas: usize = 5;
    const nshells: usize = 7;
    
    let mut atm: [i32; natm * ATM_SLOTS] = [8, 20, 1, 23, 0, 0, 
                                                1, 24, 1, 27, 0, 0, 
                                                1, 28, 1, 31, 0, 0];
    let mut bas: [i32; nbas * BAS_SLOTS] = [0, 0, 3, 1, 0, 38, 41, 0, 
                                                0, 0, 3, 1, 0, 44, 47, 0, 
                                                0, 1, 3, 1, 0, 50, 53, 0, 
                                                1, 0, 3, 1, 0, 32, 35, 0, 
                                                2, 0, 3, 1, 0, 32, 35, 0];
    let mut env: [f64; 56] = [0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
    -0.21042327, 0., 0., 0., 0.8416929, -1.47972415, 0., 0., 0.8416929, 1.47972415, 0., 3.42525091, 0.62391373, 0.1688554, 0.98170675,
    0.94946401, 0.29590645, 130.70932, 23.808861, 6.4436083, 15.07274649, 14.57770167, 4.54323359, 5.0331513, 1.1695961, 0.380389, -0.848697, 1.13520079,
    0.85675304, 5.0331513, 1.1695961, 0.380389, 3.42906571, 2.15628856, 0.34159239
    ];

    let mut S: Vec<f64> = int(natm, nbas, nshells, &mut atm, &mut bas, &mut env);

    print!("[");
    for i in 0..nshells {
        print!("[");
        for j in 0..nshells {
            print!("{:.5}, ", S[i*nshells + j]);
        }
        println!("], ");
    }
}